state/protocol/protocol_state/epochs/fallback_statemachine.go

```diff
@@ -196,6 +192,11 @@ func (m *FallbackStateMachine) ProcessEpochRecover(epochRecover *flow.EpochRecov
 			return false, nil
 		}
 	}
+	err = m.ejector.TrackDynamicIdentityList(m.state.NextEpoch.ActiveIdentities)
+	if err != nil {
+		m.telemetry.OnInvalidServiceEvent(epochRecover.ServiceEvent(), err)
+		return false, nil
```

Comment:
```suggestion
		if protocol.IsInvalidServiceEventError(err) {
			m.telemetry.OnInvalidServiceEvent(epochRecover.ServiceEvent(), err)
			return false, nil
		}
		return fmt.Errorf("unexpected errors tracking identity list: %w", err)
```

`TrackDynamicIdentityList` doesn't return anything besides `InvalidServiceEventError` now, but it's better to handle the non-sentinel case upfront to be more resilient to future changes.

Reply:
I am not sure how to approach it. Previously we had similar problem in other helper functions in FallbackStateMachine, I have end up relying on documentation that states that ONLY a sentinel is possible, and didn't add any type checks for error. If I add this then it makes sense to change those cases as well. Maybe we should discuss it?

Reply:
> similar problem in other helper functions in FallbackStateMachine, I have end up relying on documentation that states that ONLY a sentinel is possible

From my perspective, your code _largely_ follows our [Coding Convention](https://github.com/onflow/flow-go/blob/aa5cd6ac03601533604280dfb79cf356dcac4bec/CodingConventions.md#best-practice-guidelines) 👉 section `Optional Simplification for components that solely return benign errors`. Therefore, I think it is up to you as an implementor. However, for the _current_ `ejector` implementation (and largely only there), I personally agree with Jordan's suggestion.

I'd like to offer the following thoughts:
* For side-effect-free functions that can only return benign errors, it is totally fine to not check error types. Your function `IsValidExtendingEpochCommit` is a good example: https://github.com/onflow/flow-go/blob/3e7474583fa6f455c6a953d870bb6a25d7fb18a5/state/protocol/validity.go#L133-L134
  * From the signature, it is clear that the function does not operate on _any_ stateful objects, despite the input. It being a pure function greatly reduces that changes that after a refactoring, exceptions are returned.
  * The documentation is clear that the function is "side-effect free" and that it can _only_ return errors of one specific type, which have a reasonably intuitive meaning.
* Lets now take a look at a similar method: https://github.com/onflow/flow-go/blob/3e7474583fa6f455c6a953d870bb6a25d7fb18a5/state/protocol/validity.go#L153-L155 
  When reading the implementation (ideally something that shouldn't be necessary to use the function correctly) it is clear that the function is also a side-effect-free. Though, in my opinion, the documentation should be improved:
   - Currently it states "Expected errors during normal operations", which implies that there _could_ be exceptions returned.
   - Furthermore, I would appreciate the documentation emphasizing the function's "side-effect-free" nature (it doesn't mutate the inputs or other state, so the function erroring and leaving the inputs in a partially changed state is of no concern here)

   Nevertheless, the pattern of _not_ needing to check error types is also fine for this function. 
* Lastly, lets take a look at `ejector.TrackDynamicIdentityList`: 
  https://github.com/onflow/flow-go/blob/3e7474583fa6f455c6a953d870bb6a25d7fb18a5/state/protocol/protocol_state/epochs/identity_ejector.go#L50-L54
   * Note that this method is operating on a _stateful object_. Hence, accidentally changing the implementation so that it leaves the ejector in a broken state is actually quite plausible. 
   * On the happy path, this method is _not_ side-effect free (on purpose): its core functionality is to change the receiver!

Therefore, I think `TrackDynamicIdentityList`  is meaningfully different to challenge the simplified error handling. Specifically, you need _two_ conventions to be upheld for the correctness of the code:
1. only benign errors are returned _and_
2. in case of _any_ error, `TrackDynamicIdentityList` must leave the `ejector` unchanged (because we will potentially still use the ejector later)

If either is accidentally broken in the future, the revised implementation will be incorrect.


Suggestion:
1. update the documentation of [`IsValidEpochCommit`](https://github.com/onflow/flow-go/blob/3e7474583fa6f455c6a953d870bb6a25d7fb18a5/state/protocol/validity.go#L153-L155) and add
   ```
   // This is a side-effect-free function. This function only returns protocol.InvalidServiceEventError as errors.
   ```
2. for the state-full `ejector` I would indeed check the error type in the 5 different locations where you call `TrackDynamicIdentityList` (the tests already verify that we return `protocol.InvalidServiceEventError`). I think in the spirit of defensive programming, that is warranted for a stateful object.

Reply:
@AlexHentschel @jordanschalm FYI I've added checks only in two places, other cases should be treated as exceptions in my opinion. [`192c0f9` (#6168)](https://github.com/onflow/flow-go/pull/6168/commits/192c0f9d3974f4cf24997b69f9e3e88e5d24cbbb)